public/app/features/dashboard/components/Inspector/hooks.ts

```diff
@@ -0,0 +1,64 @@
+import { DataQueryError, DataSourceApi, PanelData, PanelPlugin } from '@grafana/data';
+import useAsync from 'react-use/lib/useAsync';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { DashboardModel } from 'app/features/dashboard/state';
+import { useMemo } from 'react';
+import { supportsDataQuery } from '../PanelEditor/utils';
+import { InspectTab } from './types';
+
+/**
+ * Given PanelData return first data source supporting metadata inspector
+ */
+export const useDatasourceMetadata = (data?: PanelData) => {
+  const state = useAsync<DataSourceApi | undefined>(async () => {
+    const targets = data?.request?.targets || [];
+
+    if (data && data.series && targets.length) {
+      for (const frame of data.series) {
+        if (frame.meta && frame.meta.custom) {
+          // get data source from first query
+          const dataSource = await getDataSourceSrv().get(targets[0].datasource);
+          if (dataSource && dataSource.components?.MetadataInspector) {
+            return dataSource;
+          }
+        }
+      }
+    }
+
+    return undefined;
+  }, [data]);
+  return state.value;
+};
```

Comment:
Someday we need to discuss/decide when to use Hooks, and when to use utility functions. :)

Reply:
Right. In this case, I see the use case for both. The actual logic for metadata retrieval as a utility function, and hook for providing it asynchronously to the React component. I'm gonna refactor this.